SPARK-8064, build against Hive 1.2.1

steveloughran opened the pull request:
Cherry picked the parts of the initial SPARK-8064 WiP branch needed to get sql/hive to compile against hive 1.2.1. That's the ASF release packaged under org.apache.hive, not any fork.

Tests not run yet: that's what the machines are for

steveloughran committed:
SPARK-8064, build against Hive 1.2.1

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36407 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36407/consoleFull) for   PR 7191 at commit [`9c0336c`](https://github.com/apache/spark/commit/9c0336c66985638af0ad180b0c5b2672d956a4d6).

SparkQA commented:
[Test build #36407 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36407/console) for   PR 7191 at commit [`9c0336c`](https://github.com/apache/spark/commit/9c0336c66985638af0ad180b0c5b2672d956a4d6).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

rxin commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36412 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36412/consoleFull) for   PR 7191 at commit [`9c0336c`](https://github.com/apache/spark/commit/9c0336c66985638af0ad180b0c5b2672d956a4d6).

SparkQA commented:
[Test build #36412 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36412/console) for   PR 7191 at commit [`9c0336c`](https://github.com/apache/spark/commit/9c0336c66985638af0ad180b0c5b2672d956a4d6).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

marmbrus commented:
jenkins test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37253 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37253/consoleFull) for   PR 7191 at commit [`9c0336c`](https://github.com/apache/spark/commit/9c0336c66985638af0ad180b0c5b2672d956a4d6).

SparkQA commented:
[Test build #37253 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37253/console) for   PR 7191 at commit [`9c0336c`](https://github.com/apache/spark/commit/9c0336c66985638af0ad180b0c5b2672d956a4d6).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37360 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37360/consoleFull) for   PR 7191 at commit [`b87eebf`](https://github.com/apache/spark/commit/b87eebf9e52b0e37d5e8a1fbf9c878c040060df2).

SparkQA commented:
[Test build #37360 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37360/console) for   PR 7191 at commit [`b87eebf`](https://github.com/apache/spark/commit/b87eebf9e52b0e37d5e8a1fbf9c878c040060df2).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8064 thrift server switched to Hive 1.2.1, though it doesn't compile everywhere

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37386 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37386/consoleFull) for   PR 7191 at commit [`8667845`](https://github.com/apache/spark/commit/8667845d7323e8453939e49a6575e8d76bbd6c73).

SparkQA commented:
[Test build #37386 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37386/console) for   PR 7191 at commit [`8667845`](https://github.com/apache/spark/commit/8667845d7323e8453939e49a6575e8d76bbd6c73).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8064 subclass  Hive ServerOptionsProcessor to make it public again

steveloughran committed:
SPARK-8064 everything compiles with some commenting and moving of classes into a hive package

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37538 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37538/consoleFull) for   PR 7191 at commit [`4f1e210`](https://github.com/apache/spark/commit/4f1e21065d95386c5e1ffd026fc252dbda0168c5).

SparkQA commented:
[Test build #37538 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37538/console) for   PR 7191 at commit [`4f1e210`](https://github.com/apache/spark/commit/4f1e21065d95386c5e1ffd026fc252dbda0168c5).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class HiveServerServerOptionsProcessor(serverName: String)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8054 cut pentaho dependency from calcite

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37649 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37649/consoleFull) for   PR 7191 at commit [`778e97e`](https://github.com/apache/spark/commit/778e97e59420a0290c394e1ab380ca17d10ce9c2).

SparkQA commented:
**[Test build #37649 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37649/console)**     for PR 7191 at commit [`778e97e`](https://github.com/apache/spark/commit/778e97e59420a0290c394e1ab380ca17d10ce9c2)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38105 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38105/consoleFull) for   PR 7191 at commit [`7d0fcf1`](https://github.com/apache/spark/commit/7d0fcf156d9302ba3c735d6c847e863827b75b3f).

steveloughran committed:
SPARK-8064 stop excluding hive shims from dependency (commented out , for now)

steveloughran committed:
SPARK-8064 fix some style & IDE warnings

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

steveloughran committed:
SPARK-8064: SparkSQLCLIDriver drops remote mode support. CLISuite Tests pass instead of timing out: undetected regression?

SparkQA commented:
[Test build #38107 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38107/consoleFull) for   PR 7191 at commit [`9772354`](https://github.com/apache/spark/commit/977235454b90fa7795b4540b5a745ebacf72d72a).

SparkQA commented:
[Test build #38107 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38107/console) for   PR 7191 at commit [`9772354`](https://github.com/apache/spark/commit/977235454b90fa7795b4540b5a745ebacf72d72a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateArray(children: Seq[Expression]) extends Expression `
  * `case class CreateStruct(children: Seq[Expression]) extends Expression `
  * `case class CreateNamedStruct(children: Seq[Expression]) extends Expression `

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8064 move strings used to check for http/bin thrift services up into constants

steveloughran committed:
SPARK-8064 change thrift binary startup message to "Starting ThriftBinaryCLIService on port"

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38117 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38117/consoleFull) for   PR 7191 at commit [`ced5c09`](https://github.com/apache/spark/commit/ced5c097eca03c7f0eb48fdf5cc0913eec57d04c).

SparkQA commented:
**[Test build #38105 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38105/console)**     for PR 7191 at commit [`7d0fcf1`](https://github.com/apache/spark/commit/7d0fcf156d9302ba3c735d6c847e863827b75b3f)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
**[Test build #38117 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38117/console)**     for PR 7191 at commit [`ced5c09`](https://github.com/apache/spark/commit/ced5c097eca03c7f0eb48fdf5cc0913eec57d04c)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8064 (unsuccessful) attempt to get rid of pentaho as a transitive dependency of hive-exec

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38253 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38253/consoleFull) for   PR 7191 at commit [`c7d8d82`](https://github.com/apache/spark/commit/c7d8d82a38d93220b277eed416132d33a3d4713d).

steveloughran committed:
SPARK-8064 successful attempt to get rid of pentaho as a transitive dependency of hive-exec

steveloughran committed:
SPARK-8064 run-tests.py adds a hadoop-2.6 profile, and changes info messages to say "w/Hive 1.2.1" in console output

steveloughran committed:
SPARK-8064: when a CLI suite test fails include the full output text in the raised exception; this ensures that the stdout/stderr is included in jenkins reports, so it becomes possible to diagnose the cause.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38254 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38254/consoleFull) for   PR 7191 at commit [`01e2bde`](https://github.com/apache/spark/commit/01e2bdea79a510c2fd46ae0c936fd1a136587a5d).

vanzin commented:
Hi @steveloughran,

I actually spent some time recently updating the Hive code to work against 1.1 (not 1.2, but a lot of it should apply). I can't post it as a PR because I did not touch the thrift server code, but since you're working on that part also, I thought you'd benefit from the code I worked on.

Looking at your patch I see a lot of things that I had to change missing. I've posted my patch here:
https://github.com/vanzin/spark/tree/hive-1.1

Feel free to use it any way you see fit. All the `sql/hive` unit tests pass with that patch, when built with Hive 1.1.

SparkQA commented:
**[Test build #38253 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38253/console)**     for PR 7191 at commit [`c7d8d82`](https://github.com/apache/spark/commit/c7d8d82a38d93220b277eed416132d33a3d4713d)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8064: test resilience/assertion improvements. Independent of the rest of the work; can be backported to earlier versions

steveloughran committed:
SPARK-8064: name of unrecognized file format wasn't appearing in error text

steveloughran commented:
vanzin: thanks for that work. I'd held off going near sql/hive as you'd said you were working on it, but I was just starting to stare at the tokenization code and wondering what I was going to do there...

I'm going to push up my next iteration, which adds more resilience to the tests (i.e. cleanup), but doesn't address any of the root causes.  Then I'll see about how to merge in your code.

vanzin commented:
>  I'd held off going near sql/hive as you'd said you were working on it

I was working on the metastore support for newer Hive versions, which went in some time ago. Sorry if that wasn't clear and you've been holding back on your work for that reason - the patch I posted above is unrelated to the metastore work.

SparkQA commented:
**[Test build #38254 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38254/console)**     for PR 7191 at commit [`01e2bde`](https://github.com/apache/spark/commit/01e2bdea79a510c2fd46ae0c936fd1a136587a5d)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8064: delicate merge in of the branch vanzin/hive-1.1

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38301 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38301/consoleFull) for   PR 7191 at commit [`c303d2f`](https://github.com/apache/spark/commit/c303d2f7173a4b062be2dc73340bd809b1f25ac2).

SparkQA commented:
[Test build #38301 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38301/console) for   PR 7191 at commit [`c303d2f`](https://github.com/apache/spark/commit/c303d2f7173a4b062be2dc73340bd809b1f25ac2).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnionType(types: Seq[DataType]) extends DataType `

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8064 improve diagns on uknown token; fix scalastyle failure

steveloughran committed:
SPARK-8064: HIVE-9039 renamed TOK_UNION => TOK_UNIONALL while adding TOK_UNIONDISTINCT

steveloughran committed:
SPARK-8064: DATE is now a predicate; you can't use it as a field in select ops

steveloughran committed:
SPARK-8064 Hive compatibilty tests sin sync with explain/show output from Hive 1.2.1

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38396 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38396/consoleFull) for   PR 7191 at commit [`1eeb87f`](https://github.com/apache/spark/commit/1eeb87f6bdf152bc2e2f8b30266ea581c97bfa3c).

steveloughran committed:
SPARK-8064: thrift server startup to fail fast on any exception in the main thread

steveloughran committed:
SPARK-8064: reinstate yarn-rm-server dependencies to hive-exec to ensure that jersey server is on classpath on hadoop versions < 2.6

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
**[Test build #38396 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38396/console)**     for PR 7191 at commit [`1eeb87f`](https://github.com/apache/spark/commit/1eeb87f6bdf152bc2e2f8b30266ea581c97bfa3c)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran commented:
Test build 38301 timed out, presumably due to timeout on thrift server tests that took 6+ minutes to fail if the server didn't come up. Changes to the pom  to get jersey server back on the classpath on hadoop < 2.6, (i.e. reinstate hive-exec's inclusion of the yarn-rm-server module) and improved fail-fast test setup should address this.

as a result of the timeout, failures in the `HiveCompatibilitySuite` haven't been picked up

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

steveloughran committed:
SPARK-8064: more resubmit than fix; tighten startup timeout to 60s. Still no obvious reason why jersey server code in spark-assembly isn't being picked up -it hasn't been shaded

SparkQA commented:
[Test build #38508 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38508/consoleFull) for   PR 7191 at commit [`a1907bf`](https://github.com/apache/spark/commit/a1907bf198bedeae7aa60a75bde54a45832eb894).

SparkQA commented:
[Test build #38508 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38508/console) for   PR 7191 at commit [`a1907bf`](https://github.com/apache/spark/commit/a1907bf198bedeae7aa60a75bde54a45832eb894).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnionType(types: Seq[DataType]) extends DataType `

AmplabJenkins commented:
Build finished. Test FAILed.

steveloughran committed:
SPARK-8064 TestHive to use sered2.thrift.test.Complex

steveloughran committed:
SPARK-8064 ParquetSuites to escape `date` field name

steveloughran committed:
SPARK-8064 wrap tests with withTable() calls to avoid table-exists exceptions

steveloughran committed:
SPARK-8064 missed one of the `date` escapes

steveloughran committed:
SPARK-8064 rename field double to  dbl

steveloughran committed:
SPARK-8064 fail fast on thrive process spawning tests on exit codes and/or error string patterns seen in log.

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

AmplabJenkins commented:
Build finished. Test FAILed.

steveloughran committed:
SPARK-8064 add extra logging to the YarnClusterSuite classpath test. There should be no reason why this is failing on jenkins, but as it is (and presumably its CP-related), improve the logging including any exception raised.

steveloughran committed:
SPARK-8064 reinstate something cut briefly during rebasing

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38621 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38621/consoleFull) for   PR 7191 at commit [`d7553b0`](https://github.com/apache/spark/commit/d7553b059bf5a2be894f8657a04f69da126c36c8).

steveloughran committed:
SPARK-8064 add explicit reason for kv split failure, rather than array OOB. *does not address the issue*

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38647 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38647/consoleFull) for   PR 7191 at commit [`c2ca071`](https://github.com/apache/spark/commit/c2ca071e4d17ddd6352d51d134c9c9b8cc6c25bb).

SparkQA commented:
[Test build #38621 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38621/console) for   PR 7191 at commit [`d7553b0`](https://github.com/apache/spark/commit/d7553b059bf5a2be894f8657a04f69da126c36c8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RFormula(override val uid: String) extends Estimator[RFormulaModel] with RFormulaBase `
  * `case class UnionType(types: Seq[DataType]) extends DataType `

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #38647 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38647/console) for   PR 7191 at commit [`c2ca071`](https://github.com/apache/spark/commit/c2ca071e4d17ddd6352d51d134c9c9b8cc6c25bb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnionType(types: Seq[DataType]) extends DataType `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

steveloughran committed:
SPARK-8064: on failures executing spark-submit in HiveSparkSubmitSuite, print command line and all logged output.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

pwendell commented:
Hey @steveloughran and @vanzin. If I understand correctly, I think we need to modify the dependency on hive-exec to depend on the `core` artifact. Currently I think this depends on the default `hive-exec` artifact, which is a 20+MB jar that includes a bunch of other dependencies.

http://search.maven.org/#artifactdetails%7Corg.apache.hive%7Chive-exec%7C1.2.1%7Cjar

vanzin commented:
I'm not super familiar with hive's artifacts, but if we can avoid depending on hive-exec that would be great.

pwendell commented:
Yeah the issue is that the `hive-exec` jar in upstream hive is actually an assembly jar that includes guava and a bunch of other things. This causes dependency conflicts that are basically impossible to fix, since those packages aren't isolated to their own jar.

https://issues.apache.org/jira/browse/HIVE-5733

They fixed this in Hive 1.2, by adding a new `hive-exec` under a classifier "core" that doesn't include them. We need to make sure we are using that variant of hive-exec. Currently Spark depends on a forked version of hive 13 that modifies the hive-exec jar to not be an assembly jar.

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/types/UnionType.scala:
```diff
@@ -0,0 +1,51 @@
+import org.apache.spark.annotation.DeveloperApi
+
+@DeveloperApi
+case class UnionType(types: Seq[DataType]) extends DataType {
```
As we discussed offline, I think it would be better if we could break the addition of union types off into a different PR.

vanzin replied:
Speaking as the person who wrote this code - I'm not against removing it from this PR, just note that one test doesn't work without this code ("select from thrift based table" in HiveQuerySuite). Which I think means that any thrift-based tables won't work without it.

marmbrus replied:
I see.  Instead of adding union type, could we just handle Union types with a single element? (is that is what is happening in that case?)

vanzin replied:
Not really. From what I remember, the metadata for the table contains a properties map where the value is a union of all other types, so it's not just a single element (if I understand what you mean).

marmbrus replied:
I see, thanks for clarifying.  I think its okay to simplify that test to use only atomic types in that case.

vanzin replied:
Hmm, note that the problem is not with the contents of the table, but with the table metadata.

 (In my original patch I forked Hive's `ComplexType.java` from an old version that didn't include any union columns. But I still needed `UnionType` to parse the table metadata in the first place.)

vanzin replied:
Hmm... after looking at the hive code again, maybe I got confused in my testing. Forking the old version of `org/apache/hadoop/hive/serde2/thrift/test/Complex.java` might be enough to avoid needing to add support for union types.

(Would be nice to test, though, to make sure. I'll check in my local branch.)

marmbrus replied:
Sure but we can either create a table/test with a simpler schema than `org.apache.hadoop.hive.serde2.thrift.test.Complex`  right?  Either that or we can just point users at spark-avro.

vanzin replied:
> we can either create a table/test with a simpler schema

Yes, but isn't the goal here to test support for the thrift serdes? Writing a non-thrift type defeats that. And writing a different thrift type means you'd have to generate new data (`sql/hive/src/test/resources/data/files/complex.seq`). Not an insurmountable problem, but forking `Complex.java` from Hive 0.13 is not horrible either.

marmbrus replied:
Sorry, I don't know how i got thrift/avro mixed up... duh

If there is any easy path where we just fork `Complex.java` that sounds good to me too.

vanzin replied:
Just ran HiveQueryTest without `UnionType` and it passes. All that is needed is to fork `Complex.java` and change `TestHive.scala` to use it (I changed the package name in my fork).

marmbrus replied:
Awesome!

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataTypeParser.scala:
```diff
@@ -83,10 +83,16 @@ private[sql] trait DataTypeParser extends StandardTokenParsers {
     arrayType |
     mapType |
     structType |
+    unionType |
```
Same here.

marmbrus commented on sql/catalyst/src/test/scala/org/apache/spark/sql/types/DataTypeParserSuite.scala:
```diff
@@ -110,6 +110,11 @@ class DataTypeParserSuite extends SparkFunSuite {
+  checkDataType(
+    """uniontype<int,bigint,string,double,boolean,array<string>,map<string,string>>""",
+    UnionType(Array(IntegerType, LongType, StringType, DoubleType, BooleanType,
+      ArrayType(StringType), MapType(StringType, StringType))))
```
remove also

marmbrus commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala:
```diff
@@ -40,6 +41,10 @@ import org.apache.spark.Logging
 
+/**
+ * This code doesn't support remote connections in Hive 1.2+, as the underlying CliDriver
+ * has dropped its support.
```
Maybe just document what is supported and what this class is for.

marmbrus commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala:
```diff
@@ -239,25 +251,33 @@ private[hive] class SparkSQLCLIDriver extends CliDriver with Logging {
     SparkSQLEnv.init()
+  } else {
+    // Hive 1.2 + not supported in CLI
+    throw new RuntimeException("Remote operations not supported")
```
Here I'd say something like "Starting with Spark 1.5 the Spark SQL CLI cannot be used to execute remote operations.  Instead, start the the Spark SQL thrift server and run commands using bin/beeline".

marmbrus commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala:
```diff
@@ -131,6 +129,9 @@ private[hive] object SparkSQLCLIDriver extends Logging {
       Thread.currentThread().setContextClassLoader(loader)
+    } else {
+      // Hive 1.2 + not supported in CLI
+      throw new RuntimeException("Remote operations not supported")
```
Same error with a suggested resolution here.

marmbrus commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala:
```diff
@@ -230,6 +235,13 @@ private[hive] object SparkSQLCLIDriver extends Logging {
+
+
+  def isRemoteMode(state: CliSessionState): Boolean = {
+    //    sessionState.isRemoteMode
```
remove

marmbrus commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala:
```diff
@@ -190,10 +191,14 @@ private[hive] object SparkSQLCLIDriver extends Logging {
 
     transport = clientTransportTSocketField.get(sessionState).asInstanceOf[TSocket]
+*/
+    transport = null
```
What is going on here?

steveloughran replied:
1. There's some changes to catch method/class renaming.
2. The commented out stuff is a comment out of some code that was using reflection to play with transports. I think that could just be cut altogether, given that remote cli isn't going to be viable.

marmbrus replied:
Okay, lets remove it then.

marmbrus commented on sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetCompatibilityTest.scala:
```diff
@@ -31,6 +31,8 @@ import org.apache.spark.util.Utils
 abstract class ParquetCompatibilityTest extends QueryTest with ParquetTest with BeforeAndAfterAll {
   protected var parquetStore: File = _
 
+  protected def stagingDir: Option[String] = None
```
Can you add a comment about what this is for?

steveloughran replied:
done. Hive now uses a staging subdir (`.hive-staging`) under tables, so work there stays in the same HDFS encryption zone as the data itself...this property ensures that if the path is set, parquet files under it are ignored.

marmbrus commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:
```diff
@@ -81,6 +83,12 @@ class CliSuite extends SparkFunSuite with BeforeAndAfter with Logging {
           foundAllExpectedAnswers.trySuccess(())
         }
+      } else {
+        errorResponses.map( r => {
```
`foreach`

marmbrus commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:
```diff
@@ -50,7 +51,8 @@ class CliSuite extends SparkFunSuite with BeforeAndAfter with Logging {
       timeout: FiniteDuration,
-      extraArgs: Seq[String] = Seq.empty)(
+      extraArgs: Seq[String] = Seq.empty,
+      errorResponses: Seq[String] = Seq("Error:"))(
```
can you add scaladoc for this parameter?

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:
```diff
@@ -252,6 +252,10 @@ class HiveContext(sc: SparkContext) extends SQLContext(sc) with Logging {
   }
 
   protected[sql] override def parseSql(sql: String): LogicalPlan = {
+    var state = SessionState.get()
```
`val` or inline.

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:
```diff
@@ -515,16 +530,18 @@ class HiveContext(sc: SparkContext) extends SQLContext(sc) with Logging {
+       s" Use Hive ${hiveExecutionVersion}, which is bundled with the Spark assembly jar when " +
+      s"<code>-Phive</code> is enabled. When this option is chosen, " +
+      s"spark.sql.hive.metastore.version must be either <code>${hiveExecutionVersion}</code>" +
+        s" or not defined. " +
```
nit: can you reformat this?  maybe even use `.stripMargin`?

steveloughran replied:
I'll use stripMargin

marmbrus commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/parquetSuites.scala:
```diff
@@ -255,118 +263,118 @@ class ParquetMetastoreSuite extends ParquetPartitioningTest {
-      }.size
+      assertResult(2) {
+        analyzed.collect {
+          case r@LogicalRelation(_: ParquetRelation) => r
```
nit: spaces around `@`

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -554,27 +554,33 @@ object PushPredicateThroughProject extends Rule[LogicalPlan] with PredicateHelpe
+  private def hasNondeterministic(
+      condition: Expression,
+      sourceAliases: AttributeMap[Expression]) = {
+    condition.collect {
```
I don't think we have to do this anymore.  `deterministic` traverses the whole tree now.

marmbrus commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala:
```diff
@@ -545,23 +558,26 @@ abstract class HiveThriftServer2Test extends SparkFunSuite with BeforeAndAfterAl
-            // because of port contention).
-            serverStarted.tryFailure(new RuntimeException("Failed to start HiveThriftServer2"))
-          }
+          successLines.map(r => {
```
`foreach`

marmbrus commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala:
```diff
@@ -545,23 +558,26 @@ abstract class HiveThriftServer2Test extends SparkFunSuite with BeforeAndAfterAl
+              serverStarted.trySuccess(())
+            }
+          })
+          failureLines.map(r => {
```
`foreach`

marmbrus commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLSessionManager.scala:
```diff
@@ -55,12 +56,14 @@ private[hive] class SparkSQLSessionManager(hiveContext: HiveContext)
     hiveContext.openSession()
-    val sessionHandle = super.openSession(
-      protocol, username, passwd, sessionConf, withImpersonation, delegationToken)
+    val sessionHandle = super
```
nit: wrap super too.

steveloughran replied:
done. It's still a bit choppy
```
    val sessionHandle = 
      super.openSession(protocol, username, passwd, ipAddress, sessionConf, withImpersonation,
          delegationToken)
```
.. should I put every argument on to its own line?

marmbrus replied:
Yeah I think technically the style guide says that once you wrap you must wrap all of them.  I'm not super strict about that one though.

marmbrus commented on sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala:
```diff
@@ -239,25 +251,33 @@ private[hive] class SparkSQLCLIDriver extends CliDriver with Logging {
       cmd_trimmed.startsWith("!") ||
       tokens(0).toLowerCase.equals("list") ||
-      sessionState.isRemoteMode) {
+      isRemoteMode) {
```
is it still possible for us to be in remote mode here? or will we have already thrown an error?

marmbrus commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -554,27 +554,33 @@ object PushPredicateThroughProject extends Rule[LogicalPlan] with PredicateHelpe
       } else {
         // If they are all nondeterministic conditions, leave it un-changed.
-        if (deterministic.isEmpty) {
+        if (nondeterministicConditions.length == andConditions.length) {
```
Consider just making this an `else if` instead of nesting.

steveloughran committed:
SPARK-8064 better takeup of asynchronously caught error text

liancheng committed:
WIP: Partial fix for Thrift server and CLI tests

steveloughran committed:
SPARK-8064  more robust teardown on JavaMetastoreDatasourcesSuite

steveloughran committed:
SPARK-8064 address review comments

steveloughran committed:
SPARK-8064 ignore RTE raised in blocking process.exitValue() call

steveloughran committed:
SPARK-8064 switch to an unshaded version of hive-exec-core, which must have been built with Kryo 2.21. This currently looks for a (locally built) version 1.2.1.spark

steveloughran committed:
SPARK-8064 purge UnionType

steveloughran committed:
SPARK-8064 revert to master's determinstic pushdown strategy

steveloughran committed:
SPARK-8064 version of hive to d/l from ivy is 1.2.1

steveloughran committed:
SPARK-8064 poms rely implicitly on the version of kryo chill provides

steveloughran committed:
HIVE-8064 use subset of hive complex type whose types deserialize

steveloughran committed:
SPARK-8064 fix compilation regression surfacing from rebase

steveloughran committed:
HIVE-8064 handle raw and escaped property tokens

steveloughran committed:
SPARK-8064 rely on unescaping to have already taken place, so go straight to map of serde options

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38928 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38928/consoleFull) for   PR 7191 at commit [`dbcc6a6`](https://github.com/apache/spark/commit/dbcc6a60f8c8b0dd250c2baf47ba019cf4a9654a).

SparkQA commented:
[Test build #38928 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38928/console) for   PR 7191 at commit [`dbcc6a6`](https://github.com/apache/spark/commit/dbcc6a60f8c8b0dd250c2baf47ba019cf4a9654a).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8064 IDE-hinted cleanups of Complex.java to reduce compiler warnings. It's all autogenerated code, so still ugly.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38986 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38986/consoleFull) for   PR 7191 at commit [`fe631e3`](https://github.com/apache/spark/commit/fe631e3feaffac7d5e64fc7f4b3d48e29c78a481).

SparkQA commented:
[Test build #38986 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38986/console) for   PR 7191 at commit [`fe631e3`](https://github.com/apache/spark/commit/fe631e3feaffac7d5e64fc7f4b3d48e29c78a481).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8064 deprecation warning  noise in one of the tests

steveloughran commented:
We are getting build failures as I've added a dependency on a hive-1.2.1.spark JAR which needs to be build locally (for now), one that reverts back to the chill-compatible Kryo 2.21 release.

see :[steveloughran/hive/tree/stevel/branch-1.2.1.spark](https://github.com/steveloughran/hive/tree/stevel/branch-1.2.1.spark)

This will need to be published on maven central under another group ID (as it won't come out as an ASF one)

pwendell commented:
Thanks @steveloughran I can take a crack at publishing to maven. Since that might take a day or so, one thing you can do is just put the forked hive jars in your people.apache.org web space and then add that as a repository to the build (a maven repository is just anything that can support HTTP downloading of the jars). In the mean time I can try to get publishing up and running.

vanzin commented on dev/run-tests.py:
```diff
@@ -266,6 +266,7 @@ def get_hadoop_profiles(hadoop_version):
         "hadoop2.0": ["-Phadoop-1", "-Dhadoop.version=2.0.0-mr1-cdh4.1.1"],
         "hadoop2.2": ["-Pyarn", "-Phadoop-2.2"],
         "hadoop2.3": ["-Pyarn", "-Phadoop-2.3", "-Dhadoop.version=2.3.0"],
+        "hadoop2.6": ["-Pyarn", "-Phadoop-2.6", "-Dhadoop.version=2.6.0"],
```
minor nit, and I see you're just following the pattern, but the profile should already set `hadoop.version`, so re-stating it here is redundant.

vanzin commented on pom.xml:
```diff
@@ -1273,6 +1375,8 @@
               <spark.driver.allowMultipleContexts>true</spark.driver.allowMultipleContexts>
               <spark.unsafe.exceptionOnMemoryLeak>true</spark.unsafe.exceptionOnMemoryLeak>
+              <!-- Needed by sql/hive tests. -->
+              <test.src.tables>src</test.src.tables>
```
Ah, this is from my patch, but it should be the name of a table that is not used by the Spark tests (such as `__not_used__`). I guess it works because we don't have JUnit-based tests that trigger the relevant code paths.

steveloughran replied:
so I should change it to to `not_used` then?

vanzin replied:
I'd prefer that, since there are tests that actually use a table called `src`.

vanzin commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:
```diff
@@ -81,6 +94,12 @@ class CliSuite extends SparkFunSuite with BeforeAndAfter with Logging {
           foundAllExpectedAnswers.trySuccess(())
         }
+      } else {
+        errorResponses.foreach( r => {
```
nit: style is generally `.foreach { r =>`

vanzin commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:
```diff
@@ -88,16 +107,44 @@ class CliSuite extends SparkFunSuite with BeforeAndAfter with Logging {
+
+      override def run(): Unit = {
+        try {
+          exitValue = process.exitValue()
```
Don't you mean `waitFor` here? Otherwise, if the thread runs before the process has exited, you'll get an `IllegalThreadStateException`.

vanzin commented on sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala:
```diff
@@ -545,23 +558,26 @@ abstract class HiveThriftServer2Test extends SparkFunSuite with BeforeAndAfterAl
-            // because of port contention).
-            serverStarted.tryFailure(new RuntimeException("Failed to start HiveThriftServer2"))
-          }
+          successLines.foreach(r => {
```
same style nit, also in other places:

    blah.foreach { r =>
      // code
    }

vanzin commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala:
```diff
@@ -751,7 +763,7 @@ https://cwiki.apache.org/confluence/display/Hive/Enhanced+Aggregation%2C+Cube%2C
         case Token("TOK_TABLEPROPERTIES", list :: Nil) =>
           tableDesc = tableDesc.copy(properties = tableDesc.properties ++ getProperties(list))
-        case list @ Token("TOK_TABLEFILEFORMAT", _) =>
+        case list @ Token("TOK_TABLEFILEFORMAT", children) =>
```
Also leftover from my debugging sessions, this change is not really needed.

vanzin commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala:
```diff
@@ -1251,7 +1265,8 @@ https://cwiki.apache.org/confluence/display/Hive/Enhanced+Aggregation%2C+Cube%2C
     case a: ASTNode =>
-      throw new NotImplementedError(s"No parse rules for:\n ${dumpTree(a).toString} ")
+      throw new NotImplementedError(s"No parse rules for ${a.getName}:" +
+          s"\n ${dumpTree(a).toString} ")
```
not your code, but `toString` looks redundant here. (Then maybe it all fits in one line?)

vanzin commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala:
```diff
@@ -20,6 +20,8 @@ package org.apache.spark.sql.hive
 import java.io.{InputStream, OutputStream}
 import java.rmi.server.UID
 
+import org.apache.avro.Schema
```
super nit: import ordering.

vanzin commented:
Commented on a few nits plus a couple of minor test things, but looks ok to me. I'll leave it to Michael, though, who is more familiar with the whole feature, to have a final say.

steveloughran committed:
SPARK-8064 manage the Hive depencencies to that
-things that aren't needed are excluded
-sql/hive built with ivy is in sync with the maven reconciliation policy, rather than latest-first

steveloughran committed:
HIVE-8064 fix up hive-thriftserver dependencies and cut back on evicted references in the hive- packages; this keeps mvn and ivy resolution compatible, as the reconciliation policy is "by hand"

steveloughran committed:
SPARK-8064  maven/ivy dependency purge; calcite declaration needed

steveloughran committed:
SPARK-8064 try to supress compiler warnings on Complex.java pasted-thrift-code

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39142 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39142/consoleFull) for   PR 7191 at commit [`de4447a`](https://github.com/apache/spark/commit/de4447a33281ae7ff6491cea97a53559b228a24d).

SparkQA commented:
[Test build #39142 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39142/console) for   PR 7191 at commit [`de4447a`](https://github.com/apache/spark/commit/de4447a33281ae7ff6491cea97a53559b228a24d).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

pwendell committed:
Add spark-hive staging repo

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39157 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39157/consoleFull) for   PR 7191 at commit [`baed26e`](https://github.com/apache/spark/commit/baed26e6e932ca6ddd053792d134af1baa0759b7).

SparkQA commented:
[Test build #39157 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39157/console) for   PR 7191 at commit [`baed26e`](https://github.com/apache/spark/commit/baed26e6e932ca6ddd053792d134af1baa0759b7).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

marmbrus committed:
Fixes for Hive tests.

steveloughran committed:
SPARK-8064 cherry-pick-incomplete

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39165 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39165/consoleFull) for   PR 7191 at commit [`7ffb844`](https://github.com/apache/spark/commit/7ffb84424f43caf7e5bbce51a98d299485498140).

SparkQA commented:
[Test build #39165 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39165/console) for   PR 7191 at commit [`7ffb844`](https://github.com/apache/spark/commit/7ffb84424f43caf7e5bbce51a98d299485498140).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8064 explicit protobuf import

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39268 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39268/consoleFull) for   PR 7191 at commit [`6d5fdd5`](https://github.com/apache/spark/commit/6d5fdd59a4a75f44edc16b138299a931d43f590a).

steveloughran committed:
SPARK-8064 remove hadoop.version from runtest.py, as profile will fix that automatically.

SparkQA commented:
[Test build #39268 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39268/console) for   PR 7191 at commit [`6d5fdd5`](https://github.com/apache/spark/commit/6d5fdd59a4a75f44edc16b138299a931d43f590a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8064 switch to the protobuf shaded hive-exec jar with tests to chase it down

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #39323 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39323/consoleFull) for   PR 7191 at commit [`d875543`](https://github.com/apache/spark/commit/d875543b18d0a2074544745537133ba8bd014ca0).

SparkQA commented:
[Test build #39323 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39323/console) for   PR 7191 at commit [`d875543`](https://github.com/apache/spark/commit/d875543b18d0a2074544745537133ba8bd014ca0).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

steveloughran committed:
SPARK-8064 purge duplicate protobuf declaration

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #39327 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39327/consoleFull) for   PR 7191 at commit [`be411ea`](https://github.com/apache/spark/commit/be411ea9cb1167cda281a2f4635296a5bf93f88b).

SparkQA commented:
[Test build #39327 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39327/console) for   PR 7191 at commit [`be411ea`](https://github.com/apache/spark/commit/be411ea9cb1167cda281a2f4635296a5bf93f88b).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

steveloughran committed:
SPARK-8064 switch to second staging repo of the spark-hive artifacts. This one has the protobuf-shaded hive-exec jar

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #39365 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39365/consoleFull) for   PR 7191 at commit [`e4447b0`](https://github.com/apache/spark/commit/e4447b0430d579da7f7401818936137fe6299fb2).

SparkQA commented:
[Test build #39365 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39365/console) for   PR 7191 at commit [`e4447b0`](https://github.com/apache/spark/commit/e4447b0430d579da7f7401818936137fe6299fb2).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

steveloughran committed:
SPARK-8064 classpath dependency suite to be in sync with shading in final (?) hive-exec spark

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39368 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39368/consoleFull) for   PR 7191 at commit [`fdf759b`](https://github.com/apache/spark/commit/fdf759b4582e7e2d8edd65ca89cc7576da8f39c9).

SparkQA commented:
[Test build #39368 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39368/console) for   PR 7191 at commit [`fdf759b`](https://github.com/apache/spark/commit/fdf759b4582e7e2d8edd65ca89cc7576da8f39c9).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class StopWordsRemover(override val uid: String)`
  * `case class SortArray(base: Expression, ascendingOrder: Expression)`

steveloughran commented:
1. This is a rebased patch; apologies to anyone who has branched off it.
2. It uses our own spark-project/hive version, where hive-exec only contains the core hive packages (common, hive-exec, shims) *and a shaded protobuf*. The latter ensures that it works on the spark hadoop-1 branch. Longer term we'll work with the hive team to sort out "our differences"
3. if you've done the maven build on or since friday, you'll need to purge any local maven cached version

        rm -rf ~/.m2/repository/org/spark-project/hive/

4. If you've done the sbt build, do the same for its copies. 
5. Hint: the following finds all the various directories & jars; `rm` the directories

        find ~ | grep "hive-exec-1.2.1.spark"

6. The sql/hive test `org.apache.spark.sql.hive.ClasspathDependenciesSuite` looks for the various transitive dependencies that need to be shaded (protobuf) and those which must not be found shaded (kryo & its transitives). If this test is failing, the wrong version of something is being pulled in.

AmplabJenkins commented:
Merged build finished. Test FAILed.

steveloughran committed:
SPARK-8064 scalastyle rejects the standard Hadoop ASF license header...

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39371 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39371/consoleFull) for   PR 7191 at commit [`0bbe475`](https://github.com/apache/spark/commit/0bbe475e287885e09100eef2896b1a1a9411fdff).

pwendell commented:
I've cleared the maven/ivy cache on all the jenkins machines, so going to restart the build again. Jenkins, test this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39373 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39373/consoleFull) for   PR 7191 at commit [`0bbe475`](https://github.com/apache/spark/commit/0bbe475e287885e09100eef2896b1a1a9411fdff).

SparkQA commented:
[Test build #39371 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39371/console) for   PR 7191 at commit [`0bbe475`](https://github.com/apache/spark/commit/0bbe475e287885e09100eef2896b1a1a9411fdff).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #39373 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39373/console) for   PR 7191 at commit [`0bbe475`](https://github.com/apache/spark/commit/0bbe475e287885e09100eef2896b1a1a9411fdff).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

liancheng commented:
I'm looking into the failures caused by Parquet dependency. Hive shades Parquet classes into its own private namespace. I think we miss a shading somewhere in the updated POM.

liancheng committed:
Adds com.twitter:parquet-hadoop-bundle:1.6.0 for Hive Parquet SerDe

liancheng committed:
Overrides HiveConf time vars

steveloughran committed:
Merge commit '6a92bb09f46a04d6cd8c41bdba3ecb727ebb9030' into stevel/feature/SPARK-8064-hive-1.2-002

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

steveloughran commented:
Latest patch includes the commits from lliancheng for HiveSubmit & missing spark-hive parquet dependencies on the SBT test runs

SparkQA commented:
[Test build #39435 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39435/consoleFull) for   PR 7191 at commit [`ef4af62`](https://github.com/apache/spark/commit/ef4af62f35a64b36cee9ac0cf42749c3cc123231).

SparkQA commented:
[Test build #39435 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39435/console) for   PR 7191 at commit [`ef4af62`](https://github.com/apache/spark/commit/ef4af62f35a64b36cee9ac0cf42749c3cc123231).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

liancheng committed:
Updates .q files and corresponding golden files

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39562 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39562/consoleFull) for   PR 7191 at commit [`7556d85`](https://github.com/apache/spark/commit/7556d852bc4c525ad87f1c112c063ebdaececce3).

SparkQA commented:
[Test build #39562 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39562/console) for   PR 7191 at commit [`7556d85`](https://github.com/apache/spark/commit/7556d852bc4c525ad87f1c112c063ebdaececce3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

pwendell commented:
@steveloughran any issues preventing us from merging?

steveloughran commented:
None that I know of: I've tested the thriftserver code on `-Phadoop-1 -Dhadoop-version=1.2.1`, as well as all the way to Hadoop 2.7.1.

marmbrus commented:
Awesome, I'm going to merge to master and 1.5 then.

jameszhouyi commented:
Hi @steveloughran ,
Have you came across below errors when build the spark? Please correct me if any missing or wrong. (Build command: mvn -Pyarn -Phadoop-2.4 -Dhadoop.version=2.6.0 -Phive -Phive-thriftserver -DskipTests clean package)
....
....
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 14:37 min
[INFO] Finished at: 2015-08-04T14:27:08+08:00
[INFO] Final Memory: 65M/478M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project spark-hive_2.10: Could not resolve dependencies for project org.apache.spark:spark-hive_2.10:jar:1.5.0-SNAPSHOT: The following artifacts could not be resolved: org.spark-project.hive:hive-exec:jar:1.2.1.spark, org.spark-project.hive:hive-metastore:jar:1.2.1.spark: Failure to find org.spark-project.hive:hive-exec:jar:1.2.1.spark in http://foxcove3.sh.test.com:8081/artifactory/libs-release was cached in the local repository, resolution will not be reattempted until the update interval of libs-release has elapsed or updates are forced -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :spark-hive_2.10
